### PR TITLE
tech(testing): Add conditional type check utils

### PR DESCRIPTION
**Pull request by Duelsik**
As I am copying MIT code between repositories for the first time - I am not sure if I've put correct headers in files.

Solves https://github.com/denoland/deno_std/issues/2145
Required for https://github.com/denoland/deno_std/pull/2111

### CODE REVIEW

**Review by kt3k — COMMENTED** [review]
@Duelsik Thanks for doing this! I like the idea!

The module path `../std/testing/conditional_type_checks/mod.ts` feels a little too long to me. How about `testing/types.ts` or `testing/asserts/types.ts`? (Or any other suggestions are welcome)

**Comment by kt3k** [comment]
The way of attributing original license looks good to me.

**Comment by bartlomieju** [comment]
`testing/types.ts` seems preferable. @dsherret please take a look

File: testing/conditional_type_checks/LICENSE [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+// Copyright @dsherret and dsherret/conditional-type-checks contributors. All rights reserved. MIT license.
```
Comment (bartlomieju) [inline testing/conditional_type_checks/LICENSE:2]: Remove this file

File: testing/conditional_type_checks/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+- `IsNullable<T>` - Checks if `T` is possibly `null` or `undefined`.
+- `IsExact<T, U>` - Checks if `T` exactly matches `U`.
+- `Has<T, U>` - Checks if `T` has `U`.
+- `NotHas<T, U>` - Checks if `T` does not have `U`.
+- `IsAny<T>` - Checks if `T` is the `any` type.
+- `IsNever<T>` - Checks if `T` is the `never` type.
+- `IsUnknown<T>` - Checks if `T` is the `unknown` type.
+- More to come...
+
+## Ways to Test
```
Comment (bartlomieju) [inline testing/conditional_type_checks/README.md:32]: This information should be combined into `testing/README.md` (and appropriate chapter in https://github.com/denoland/manual should be updated)

**Review by bartlomieju — COMMENTED** [review]
I think we could also consider `testing/assert_types.ts`

**Comment by lino-levan** [comment]
@Duelsik any chance you could push this forward? If not, I'd be glad to take this on if the sentiment is still positive.

**Comment by Duelsik** [comment]
Feel free to continue this topic, I don't remember why it got stale.

**Comment by kt3k** [comment]
closing in favor of #2864 

Thanks @Duelsik for the initial work!

**Pull request closed without merging**